src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,1489 @@
+    }
+    else
+      outputDeltas.insert({ i, &layerDeltas[i] });
+
+    MakeAlias(*outputDeltas[i], layerDeltaMatrix,
+       network[currentLayer]->OutputSize(), batchSize, offset);
+    offset += network[currentLayer]->OutputSize() * batchSize;
+  }
+
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    Layer<MatType>* layer = network[currentLayer];
+
+    const std::vector<size_t>& parents = parentsList[currentLayer];
+    if (parents.size() > 1)
+    {
+      inputDeltas.insert({ i, new MatType() });
+
+      size_t inputSize = layer->InputDimensions()[0];
+      for (size_t j = 1; j < layer->InputDimensions().size(); j++)
+        inputSize *= layer->InputDimensions()[j];
+
+      MakeAlias(*inputDeltas[i], layerDeltaMatrix,
+        inputSize, batchSize, offset);
+      offset += inputSize * batchSize;
+    }
+    else
+    {
+      size_t onlyParent = sortedIndices[parents.front()];
+      inputDeltas.insert({ i, outputDeltas[onlyParent]});
+    }
+  }
+  extraDeltasAllocated = true;
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::InitializeGradientPassMemory(MatType& gradient)
+{
+  size_t gradientStart = 0;
+  for (size_t i = 0; i < sortedNetwork.size(); ++i)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    const size_t weightSize = network[currentLayer]->WeightSize();
+    MakeAlias(layerGradients[i], gradient, weightSize, 1, gradientStart);
+    gradientStart += weightSize;
+  }
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Forward(const MatType& input, MatType& results)
+{
+  CheckNetwork("DAGNetwork::Forward()", input.n_rows);
+
+  size_t lastLayer = sortedNetwork.back();
+  networkOutput.set_size(network[lastLayer]->OutputSize(), input.n_cols);
+
+  // equivalent to MultiLayer->Forward(), but with a concat.
+  if (sortedNetwork.size() > 1)
+  {
+    InitializeForwardPassMemory(input.n_cols);
+
+    size_t firstLayer = sortedNetwork.front();
+    network[firstLayer]->Forward(input, layerOutputs.front());
+
+    for (size_t i = 1; i < sortedNetwork.size(); i++)
+    {
+      size_t currentLayer = sortedNetwork[i];
+      const std::vector<size_t>& parents = parentsList[currentLayer];
+
+      // Concatenation
+      if (parents.size() > 1)
+      {
+        size_t axis = layerAxes[currentLayer];
+
+        size_t rows = 1;
+        for (size_t j = 0; j < axis; j++)
+          rows *= network[currentLayer]->InputDimensions()[j];
+
+        size_t slices = input.n_cols;
+        size_t numDims = network[currentLayer]->InputDimensions().size();
+        for (size_t j = axis + 1; j < numDims; j++)
+          slices *= network[currentLayer]->InputDimensions()[j];
```

Comment:
We calculate these (and other similar quantities) enough that I wonder if it might be worth caching them when we do `CheckGraph()` or some other setup function.